File: detox/src/android/espressoapi/DetoxAssertion.js
/**

	This code is generated.
	For more information see generation/README.md.
*/


function sanitize_matcher(matcher) {
  if (!matcher._call) {
    return matcher;
  }

  const originalMatcher = typeof matcher._call === 'function' ? matcher._call() : matcher._call;
  return originalMatcher.type ? originalMatcher.value : originalMatcher;
} 
class DetoxAssertion {
  static assertMatcher(viewInteraction, viewMatcher) {
    return {
      target: {
        type: "Class",
        value: "com.wix.detox.espresso.DetoxAssertion"
      },
      method: "assertMatcher",
      args: [{
        type: "Invocation",
        value: viewInteraction
      }, {
        type: "Invocation",
        value: sanitize_matcher(viewMatcher)
      }]
    };
  }

  static assertNotVisible(viewInteraction) {
    return {
      target: {
        type: "Class",
        value: "com.wix.detox.espresso.DetoxAssertion"
      },
      method: "assertNotVisible",
      args: [{
        type: "Invocation",
        value: viewInteraction
      }]
    };
  }

  static assertNotExists(viewInteraction) {
    return {
      target: {
        type: "Class",
        value: "com.wix.detox.espresso.DetoxAssertion"
      },
      method: "assertNotExists",
      args: [{
        type: "Invocation",
        value: viewInteraction
      }]
    };
  }

  static waitForAssertMatcher(viewInteraction, viewMatcher, timeoutSeconds) {
    if (typeof timeoutSeconds !== "number") throw new Error("timeoutSeconds should be a number, but got " + (timeoutSeconds + (" (" + (typeof timeoutSeconds + ")"))));
    return {
      target: {
        type: "Class",
        value: "com.wix.detox.espresso.DetoxAssertion"
      },
      method: "waitForAssertMatcher",
      args: [{
        type: "Invocation",
        value: viewInteraction
      }, {
        type: "Invocation",
        value: sanitize_matcher(viewMatcher)
      }, {
        type: "Double",
        value: timeoutSeconds
      }]
    };
  }

  static waitForAssertMatcherWithSearchAction(viewInteraction, viewMatcher, searchAction, searchMatcher
  ) {
    return {
      target: {
        type: "Class",
        value: "com.wix.detox.espresso.DetoxAssertion"
      },
      method: "waitForAssertMatcherWithSearchAction",
      args: [{
        type: "Invocation",
        value: viewInteraction
      }, {
        type: "Invocation",
        value: sanitize_matcher(viewMatcher)
      }, searchAction, {
        type: "Invocation",
        value: sanitize_matcher(searchMatcher
        )
      }]
    };
  }

}

module.exports = DetoxAssertion;

File: detox/android/detox/src/full/java/com/wix/detox/espresso/DetoxAssertion.java
package com.wix.detox.espresso;

import android.view.View;

import com.wix.detox.common.DetoxErrors.DetoxRuntimeException;
import com.wix.detox.common.DetoxErrors.StaleActionException;

import junit.framework.AssertionFailedError;

import org.hamcrest.Matcher;

import androidx.test.espresso.ViewAction;
import androidx.test.espresso.ViewInteraction;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.assertion.ViewAssertions.doesNotExist;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static com.wix.detox.espresso.assertion.ViewAssertions.matches;
import static org.hamcrest.Matchers.not;

/**
 * Created by simonracz on 10/07/2017.
 */

public class DetoxAssertion {

    private static final double NANOSECONDS_IN_A_SECOND = 1_000_000_000.0;

    private DetoxAssertion() {
        // This is a utility class and shouldn't be instantiated.
    }

    /**
     * Asserts the given matcher for the provided view interaction.
     */
    public static ViewInteraction assertMatcher(ViewInteraction viewInteraction, Matcher<View> viewMatcher) {
        return viewInteraction.check(matches(viewMatcher));
    }

    /**
     * Asserts that the given view interaction is not visible.
     */
    public static ViewInteraction assertNotVisible(ViewInteraction viewInteraction) {
        ViewInteraction result;
        try {
            result = viewInteraction.check(doesNotExist());
            return result;
        } catch (AssertionFailedError e) {
            result = viewInteraction.check(matches(not(isDisplayed())));
            return result;
        }
    }

    /**
     * Asserts that the given view interaction does not exist.
     */
    public static ViewInteraction assertNotExists(ViewInteraction viewInteraction) {
        return viewInteraction.check(doesNotExist());
    }

    /**
     * Waits until the provided matcher matches the view interaction or a timeout occurs.
     */
    public static void waitForAssertMatcher(final ViewInteraction viewInteraction, final Matcher<View> viewMatcher, double timeoutSeconds) {
        final long startTime = System.nanoTime();

        while (true) {
            long currentTime = System.nanoTime();
            long elapsedTime = currentTime - startTime;
            double elapsedSeconds = (double) elapsedTime / NANOSECONDS_IN_A_SECOND;
            if (elapsedSeconds >= timeoutSeconds) {
                throw new DetoxRuntimeException(
                    "" + timeoutSeconds + "sec timeout expired without matching of given matcher: " + viewMatcher);
            }

            try {
                viewInteraction.check(matches(viewMatcher));
                break;
            } catch (AssertionFailedError err) {
                UiAutomatorHelper.espressoSync(20);
            }
        }
    }

    /**
     * Continually asserts the provided matcher until a search action returns a matching view or a
     * `StaleActionException` error is thrown.
     */
    public static void waitForAssertMatcherWithSearchAction(
        final ViewInteraction viewInteraction,
        final Matcher<View> viewMatcher,
        final ViewAction searchAction,
        final Matcher<View> searchMatcher
    ) {
        while (true) {
            try {
                assertMatcher(viewInteraction, viewMatcher);
                break;
            } catch (AssertionFailedError err) {
                try {
                    onView(searchMatcher).perform(searchAction);
                } catch (StaleActionException exStaleAction) {
                    assertMatcher(viewInteraction, viewMatcher);
                    break;
                }
            }
        }
    }
}
